### PR TITLE
S3: Implement cross-account access

### PR DESCRIPTION
**Pull request by viren-nadkarni**
This PR adds tests for cross-account support to S3 for following operations:
- GetObject
- PutObject
- ListObjects

The actual implementation was done in Moto in below linked PRs. This also introduces a shared namespace for buckets.

Depends on:
- https://github.com/getmoto/moto/pull/6333
- https://github.com/localstack/localstack/pull/8410

cc: @dfangl

### CODE REVIEW

File: tests/integration/s3/test_s3.py [hunk 1]
```diff
@@ -4698,6 +4701,69 @@ def test_s3_intelligent_tier_config(self, aws_client, s3_create_bucket, snapshot
...
+            region_name=SECONDARY_TEST_AWS_REGION_NAME,
+        )
+
+    def test_shared_bucket_namespace(self, primary_client, secondary_client):
+        # Ensure that the bucket name space is shared by all accounts and regions
+        primary_client.create_bucket(Bucket="foo")
+
+        with pytest.raises(ClientError) as exc:
+            secondary_client.create_bucket(
+                Bucket="foo", CreateBucketConfiguration={"LocationConstraint": "ap-south-1"}
```
Comment (viren-nadkarni) [inline tests/integration/s3/test_s3.py:4727]:
```suggestion
                Bucket="foo", CreateBucketConfiguration={"LocationConstraint": SECONDARY_TEST_AWS_REGION_NAME}
```

**Review by bentsku — APPROVED** [review]
LGTM! I've read the change in moto, and it's not only `GetObject`, `PutObject` and `ListObjects` anymore, right, but it's effectively every single S3 operation that is now cross-account enabled?
This might actually fix #7084! 

This is very cool, I'm not sure how ACL are enforced on the `moto` side, but it should be okay. Awesome, thanks a lot for tackling this!

File: tests/integration/s3/test_s3.py [hunk 2]
```diff
@@ -4725,6 +4728,70 @@ def test_s3_intelligent_tier_config(self, aws_client, s3_create_bucket, snapshot
...
+    def test_shared_bucket_namespace(self, primary_client, secondary_client):
+        # Ensure that the bucket name space is shared by all accounts and regions
+        primary_client.create_bucket(Bucket="foo")
+
+        with pytest.raises(ClientError) as exc:
+            secondary_client.create_bucket(
+                Bucket="foo",
+                CreateBucketConfiguration={"LocationConstraint": SECONDARY_TEST_AWS_REGION_NAME},
+            )
+        exc.match("BucketAlreadyExists")
```
[the remark this answers is not in the document]
Comment (viren-nadkarni) [reply tests/integration/s3/test_s3.py:4757]: Yes, this should be fixed in Moto. I'll try to remember to sneak a fix for this in my next moto PR

**Comment by viren-nadkarni** [comment]
> I've read the change in moto, and it's not only GetObject, PutObject and ListObjects anymore, right, but it's effectively every single S3 operation that is now cross-account enabled?

It may, but this definitely needs further testing before we can claim so!

**Pull request merged**